magma/backend/mlir/hardware_module.py

```diff
@@ -138,9 +138,21 @@ def magma_type_to_mlir_type(type: Kind) -> MlirType:
         if issubclass(type.T, Bit):
             return magma_type_to_mlir_type(Bits[type.N])
         return hw.ArrayType((type.N,), magma_type_to_mlir_type(type.T))
+    # NOTE(rsetaluri): If the type is Tuple (i.e. has integer keys), then the
+    # struct type we emit should prefix the field key with an underscore
+    # (e.g. "_0") to make it a valid keyword. Note that this *must* be in sync
+    # with the translation of magma_tuple_get_op's which does the same.
+    if issubclass(type, Product):
+        fields = {
+            str(k): magma_type_to_mlir_type(t)
+            for k, t in type.field_dict.items()
+        }
+        return hw.StructType(tuple(fields.items()))
     if issubclass(type, m_Tuple):
-        fields = {str(k): magma_type_to_mlir_type(t)
-                  for k, t in type.field_dict.items()}
+        fields = {
+            f"_{i}": magma_type_to_mlir_type(t)
+            for i, t in enumerate(type.field_dict.values())
+        }
```

Comment:
How about using something like a `_tuple_key_to_str` like:

```suggestion
        def _tuple_key_to_str(k, type):
            return str(k) if issubclass(type, Product) else f"_{k}"

        fields = {_tuple_key_to_str(k, type): magma_type_to_mlir_type(t)
                  for k, t in type.field_dict.items()}
```

Reply:
:+1: